rs/xnet/payload_builder/src/lib.rs

```diff
@@ -910,6 +926,24 @@ pub fn get_msg_limit(subnet_id: SubnetId, state: &ReplicatedState) -> Option<usi
     }
 }
 
+/// Calculates an upper bound for how many messages can be included into a block based on how many
+/// signals the reverse stream will contain after inducting the stream slice.
+///
+/// This is computed such that the stream index of the last message is at most `slice_begin +
+/// 50_000`. Since `slice_begin` is used for garbage collecting signals, this ensures at most
+/// `50_000` signals in the reverse stream.
+pub fn get_signal_limit(begin: Option<&ExpectedIndices>, slice_begin: StreamIndex) -> usize {
+    // TODO: Replace the 50_000 with a constant. This should be the same constant used in the
+    // stream builder: `MAX_STREAM_MESSAGES`.
+    begin.map_or(50_000, |begin| {
```

Comment:
I have a hard time following why `begin` is optional here? Shouldn't we always know the begin of the slice we are about to put in a block?

Also looking at the code above it seems that we are passing `Some(...)` at all call sites.

Reply:
I moved the new limit all the way down [here](https://sourcegraph.com/github.com/dfinity/ic@c0b2ec6580bdaf80047b9e7b125a86eab424adff/-/blob/rs/xnet/payload_builder/src/certified_slice_pool.rs?L380), where the `Option` is unwrapped anyway.

One other advantage should be that I can test this is in isolation without an entire pool of > 10k messages in it. I have to give that a try first though.